### ISSUE TITLE
zombie exec/alloc

### ISSUE THREAD

**Issue by olgabot** [grailbio/reflow#43]
Hello, I ran [this workflow](https://github.com/czbiohub/reflow-workflows/blob/master/sourmash.rf) on [these samples](https://github.com/czbiohub/kmer-hashing/blob/master/sourmash/maca/facs/samples.csv) and while most (51,446/54,912) samples completed successfully, I'm trying to figure out what happened to the remaining 3,466 failed samples. Logs of the failed samples are here: `aws s3 ls s3://olgabot-maca/facs/sourmash_failed_samples/`

Looking at the logs (excerpted below), it seems most of the errors have to do with either `operation not supported: zombie alloc` or `operation not supported: zombie exec`, and I don't know what these mean. Since these jobs were `marked as run`, is the best option in this case to create a new `config.json` with just the failed samples and rerun the workflow with `-cache=off`?

<details>

```
 Wed 13 Jun - 19:11  ~/kmer-hashing/sourmash/maca/facs/failed_samples   origin ☊ master 6☀ 3● 
 ubuntu@olgabot-reflow  ls -1 | xargs tail | less
==> log.A10-B002010-3_39_F-1-1_S218 <==
            {{signature}} =
                . sha256:268a126b191343e2e56969da8f99588fd34b3c1e66ed6813725b6b309d38af15 26.9KiB
        profile:
            cpu mean=1.0 max=1.4
            mem mean=14.5MiB max=14.6MiB
            disk mean=0B max=0B
            tmp mean=0B max=0B
2018/06/12 06:14:44 retain 0s: keepalive 0a105711eedbf376 0s: operation not supported: zombie alloc
2018/06/12 06:14:44 marking run done after nonrecoverable error put 0a105711eedbf376 sha256:d73544f957060c3c72bb25088ea73434d5e06aa0a07ebf19e42dee3bbfeaf06f execconfig extern url s3://olgabot-maca/facs/sourmash/A10-B002010-3_39_F-1-1_S218.signature resources {}: operation not supported: zombie alloc
2018/06/12 06:14:44 run A10-B002010-3_39_F-1-1_S218: state: done error put 0a105711eedbf376 sha256:d73544f957060c3c72bb25088ea73434d5e06aa0a07ebf19e42dee3bbfeaf06f execconfig extern url s3://olgabot-maca/facs/sourmash/A10-B002010-3_39_F-1-1_S218.signature resources {}: operation not supported: zombie alloc

==> log.A10-B002421-3_39_F-1-1_S193 <==
2018/06/12 05:19:10  ->  sourmash.r2  65815baf run  intern .._39_F-1-1_S193/A10-B002421-3_39_F-1-1_S193_R2_001.fastq.gz
2018/06/12 05:19:10 sourmash.r2 65815baf /home/ubuntu/reflow-workflows/sourmash.rf:155:15:
        sha256:65815bafd3c8cdd872a754499b18158eacaecccf8d2909cd6959e43f281ad2cd
2018/06/12 05:19:10  ->  sourmash.r1  6c10daa4 run  intern .._39_F-1-1_S193/A10-B002421-3_39_F-1-1_S193_R1_001.fastq.gz
2018/06/12 05:19:10 sourmash.r1 6c10daa4 /home/ubuntu/reflow-workflows/sourmash.rf:154:15:
        sha256:6c10daa464ff5c9b702f45f669195047e6b6a7a0cf8798313b1d4b14b068525c
2018/06/12 05:19:25 retain 0s: keepalive e6679994848c3748 0s: operation not supported: zombie alloc
2018/06/12 05:19:25 marking run done after nonrecoverable error promote e6679994848c3748/65815bafd3c8cdd872a754499b18158eacaecccf8d2909cd6959e43f281ad2cd: operation not supported: zombie exec
2018/06/12 05:19:25 run A10-B002421-3_39_F-1-1_S193: state: done error promote e6679994848c3748/65815bafd3c8cdd872a754499b18158eacaecccf8d2909cd6959e43f281ad2cd: operation not supported: zombie exec
2018/06/12 16:53:50 run A10-B002421-3_39_F-1-1_S193: state: done error promote e6679994848c3748/65815bafd3c8cdd872a754499b18158eacaecccf8d2909cd6959e43f281ad2cd: operation not supported: zombie exec

==> log.A10-B002429-3_38_F-1-1_S39 <==
2018/06/12 05:57:47 accepted alloc ec2-34-220-214-175.us-west-2.compute.amazonaws.com:9000/451841d12ca4a34e
2018/06/12 05:59:54 run A10-B002429-3_38_F-1-1_S39: state: eval alloc ec2-34-220-214-175.us-west-2.compute.amazonaws.com:9000/451841d12ca4a34e
2018/06/12 05:59:54 turning caching off because assoc and repository are not configured
2018/06/12 05:59:54 evaluating with configuration: executor *client.clientAlloc transferer *repository.Manager flags cacheextern,nocache,nogc,norecomputeempty,topdown flowconfig hashv2 cachelookuptimeout 1m0s
2018/06/12 06:00:37 retain 0s: keepalive 451841d12ca4a34e 0s: operation not supported: zombie alloc
2018/06/12 06:00:37 marking run done after nonrecoverable error put 451841d12ca4a34e sha256:9ad5a3e87508a5b1ac6824d4470e86ee1ab674cb6612b18eef11fc5329b57093 execconfig intern url s3://czbiohub-maca/remux_data/170928_A00111_0068_AH3YKKDMXX/rawdata/A10-B002429-3_38_F-1-1_S39/A10-B002429-3_38_F-1-1_S39_R2_001.fastq.gz resources {}: operation not supported: zombie alloc
2018/06/12 06:00:37 run A10-B002429-3_38_F-1-1_S39: state: done error put 451841d12ca4a34e sha256:9ad5a3e87508a5b1ac6824d4470e86ee1ab674cb6612b18eef11fc5329b57093 execconfig intern url s3://czbiohub-maca/remux_data/170928_A00111_0068_AH3YKKDMXX/rawdata/A10-B002429-3_38_F-1-1_S39/A10-B002429-3_38_F-1-1_S39_R2_001.fastq.gz resources {}: operation not supported: zombie alloc
2018/06/12 16:54:04 run A10-B002429-3_38_F-1-1_S39: state: done error put 451841d12ca4a34e sha256:9ad5a3e87508a5b1ac6824d4470e86ee1ab674cb6612b18eef11fc5329b57093 execconfig intern url s3://czbiohub-maca/remux_data/170928_A00111_0068_AH3YKKDMXX/rawdata/A10-B002429-3_38_F-1-1_S39/A10-B002429-3_38_F-1-1_S39_R2_001.fastq.gz resources {}: operation not supported: zombie alloc
``` 

</details>

As a side note, the failed fastqs were *slightly* larger in size, but not substantially:

![image](https://user-images.githubusercontent.com/806256/41372756-12e15d16-6f03-11e8-9b8f-40f3a18ea2d0.png)

**Comment by olgabot** [comment]
Okay I reran these with higher resources (cpu: 1 -> 2, memory set at 4 GiB) and this seems to be running fine. Is there a way to have a more informative error that would let the user know they need to allocate more resources to each job? Relatedly, is there a way to specify an "elastic" amount of memory/disk that's dependent on the file size, e.g. 4*fastq.gz or something?

**Comment by mariusae** [comment]
Allocs (resource allocations) enter zombie state when they become unused. As an end-user, you shouldn't really be seeing them. In this case, it may be that, perhaps due to a networking event, or some sort of resource utilization spike, the evaluator reflow (i.e., the process/machine from which you run the `reflow` command) failed to keep the worker allocation alive (it requires a checkin--a "keep alive"--every few minutes), and it was collected by the worker node because the resources were needed elsewhere.

`reflow run/runbatch` does attempt to restart jobs automatically when they fail in this way, but it seems like we've overlooked some of the cases involving zombie allocs.

The solution to this should be to just invoke `reflow runbatch -retry`: it will restart only those jobs that did not yet successfully complete.

Meanwhile, I've created a bug to automatically restart zombie ops: https://github.com/grailbio/reflow/issues/47
